### PR TITLE
Add tests and do the fixes to pass the tests

### PR DESCRIPTION
**Pull request by dkurzaj**
- In the robots.txt if `User-agent: *` was above or below the specific line `User-agent: [My UA]` it would have a different behaviour although we should always jsut keep the specific rules. (tests: [rob2](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R23) [rob5](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R81))
- Even though there is no RFC, the crawl-delay value is mostly considered as being in seconds and not in miliseconds, so now we consider it as being seconds too (default value to 2 and not 2000). (test: [rob5](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R81))
- We used to overwrite the crawl-delay value with the default value if it was below the default value: it is not a right behaviour, so now we keep the crawl-delay value specified in the robots.txt even if it's lower. (test: [rob9](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R156))
- We now parse robots.txt in a case insensitive way (tests: [rob10](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R172) [rob13](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R223))
- We are now tolerate typos on the user-agent field. It can be (case is not sensitive): "user-agent", "useragent" or "user agent" (tests: [rob6](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R99) [rob12](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R205))
- We now support groups of user agents. If several "user-agent" lines follow each others, it means the rules which will follow will apply  to all of them (tests: [rob14](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R239) [rob15](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R260) [rob16](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R281))
- We used to match any user-agent rule which contained the specified user-agent to check as a substring. Now we only follow the rules which exactly match the name of the user-agent to check. (test: [rob21](https://github.com/QwantResearch/urlfactory/pull/17/files#diff-f527a91a14f0aed1152a9142cc633026R365))

### CODE REVIEW

File: src/robots.hpp [hunk 1]
```diff
@@ -138,13 +141,27 @@ class Robots {
   std::vector<UrlParser> walls;
   std::vector<UrlParser> doors;
 
+  static const int DEFAULT_CRAWL_DELAY = 2;
```
Comment (taynaud) [inline src/robots.hpp]: Can you specify the unit ? For instance DEFAULT_CRAWL_DELAY_S ?

Comment (dkurzaj) [reply src/robots.hpp]: The unit is not defined, it varies from usages.

File: tests/test_robots.cpp [hunk 2]
```diff
@@ -0,0 +1,380 @@
+#include <map>
```
Comment (taynaud) [inline tests/test_robots.cpp]: Why 2 files test_robots and robots_tests ? If it is a kind of todo, can you reflect that on the file name ?

File: tests/test_robots.cpp [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+                                Disallow: /bar.htm)";
+  urlfactory::Robots rob2("https://www.qwant.com", "Qwantify",
+                          "Mozilla/5.0 (compatible; Qwantify/Bleriot/1.1; "
+                          "+https://help.qwant.com/bot)",
+                          succeed_robot);
+  BOOST_TEST(!rob2.is_allowed("https://www.qwant.com/bar.htm"));
+  // TODO: The following should also be disallowed. Since there is no '$' at the
+  // end of the line, it is not an exact match rule. (source:
+  // https://tools.ietf.org/html/draft-rep-wg-topic-00#section-2.2.3)
+  BOOST_TEST(rob2.is_allowed("https://www.qwant.com/bar.html"));
```
Comment (taynaud) [inline tests/test_robots.cpp]: commented code is evil, but for this I would prefer that your code express the expected value but commented

File: src/robots.cpp [hunk 4]
```diff
@@ -123,53 +123,87 @@ void Robots::fetch_robots(Robots *rbt, std::string &robotstxt, long &http_code,
...
+        if (line.size() > pos + 6) {
           rbt->doors.push_back(UrlParser(line.substr(pos + 6)) + rbt->up_host);
-      } else if ((pos = line.find("Crawl-delay:")) != std::string::npos) {
-        if (line.size() > pos + 12)
-          rbt->crawl_delay = std::max(rbt->crawl_delay,
-                                      std::atoi(line.substr(pos + 12).c_str()));
+        }
+      } else if ((pos = line.find("crawl-delay:")) != std::string::npos) {
+        if (line.size() > pos + 12) {
+          rbt->crawl_delay = std::atoi(line.substr(pos + 12).c_str());
```
[the remark this answers is not in the document]
Comment (dkurzaj) [reply src/robots.cpp]: Thanks! It is now handled :)

File: tests/test_robots.cpp (hunk 3 above)
Comment (dkurzaj) [reply tests/test_robots.cpp]: Fixed :)

**Pull request merged**